File: compilation/type.cc
#include <map>

#include "type.h"
#include "context.h"

namespace cc
{
    int Type::get_size() const
    {
        switch (basic_type)
        {
            case CHAR:
            case U8:
            case I8: return 1;
            case U16:
            case I16: return 2;
            case U32:
            case I32:
            case F32: return 4;
            case U64:
            case I64:
            case F64:
            default:
                throw Exception("Invalid type passed to Type::get_size()");
        }
    }

    const Type* Type::get(Context* ctx, const std::string &name)
    {
        // Builtin types
        if (name == "char") return ctx->type<CHAR>();
        else if (name == "I8") return ctx->type<I8>();
        else if (name == "U8") return ctx->type<U8>();
        else if (name == "I16") return ctx->type<I16>();
        else if (name == "U16") return ctx->type<U16>();
        else if (name == "I32") return ctx->type<I32>();
        else if (name == "U32") return ctx->type<U32>();
        else if (name == "I64") return ctx->type<I64>();
        else if (name == "U64") return ctx->type<U64>();
        else if (name == "F32") return ctx->type<F32>();
        else if (name == "F64") return ctx->type<F64>();

        return ctx->type(name);
    }
}

File: compiler/type.py
import enum
from typing import Dict

from llvmlite.ir import types


# noinspection PyArgumentList
class Type(enum.IntEnum):
    void = enum.auto()
    i8 = enum.auto()
    u8 = enum.auto()
    i16 = enum.auto()
    u16 = enum.auto()
    i32 = enum.auto()
    u32 = enum.auto()
    i64 = enum.auto()
    u64 = enum.auto()
    f32 = enum.auto()
    f64 = enum.auto()
    boolean = enum.auto()
    string = enum.auto()

    __LLVM_CONVERSIONS__: Dict[int, 'Type'] = {
        i8: types.IntType(8),
        u8: types.IntType(8),
        i16: types.IntType(16),
        u16: types.IntType(16),
        i32: types.IntType(32),
        u32: types.IntType(32),
        i64: types.IntType(64),
        u64: types.IntType(64),
    }

    __UNSIGNED_TYPES__ = {
        u8, u16, u32, u64,
    }

    __SIGNED_TYPES__ = {
        i8, i16, i32, i64,
    }

    __FLOATING_TYPES__ = {
        f32, f64,
    }

    __INTEGRAL_TYPES__ = {
        *__SIGNED_TYPES__,
        *__UNSIGNED_TYPES__,
        *__FLOATING_TYPES__
    }

    __IMMEDIATE_TYPES__ = {
        *__INTEGRAL_TYPES__, boolean
    }

    @property
    def is_unsigned(self) -> bool:
        return self in self.__UNSIGNED_TYPES__

    @property
    def is_signed(self) -> bool:
        return self in self.__SIGNED_TYPES__

    @property
    def is_floating(self) -> bool:
        return self in self.__FLOATING_TYPES__

    @property
    def is_integer(self) -> bool:
        return self.is_unsigned or self.is_signed

    @property
    def is_integral(self) -> bool:
        return self in self.__INTEGRAL_TYPES__

    def lower(self) -> types.Type:
        """
        Get the LLVM type from the Ast type
        Note: There is no distinction between signed and unsigned
              types once the Ast type is lowered. This is the same
              as LLVM itself.
        :return: LLVM Type instance
        """
        return self.__LLVM_CONVERSIONS__[self]
